6201-120303D Латахин Егор Дмитриевич

Dakulzz opened the pull request.

Dakulzz committed:
Completed Labwork: Latakhin E.D. - 6201-120303D

Dakulzz committed:
Delete Readme.md

Dakulzz committed:
Delete assignment.md

DiBoS290699 reviewed (commented):
# Результаты проверки текста отчета
## Рецензия на отчёт по лабораторной работе №3 (ООП на Java)

**1. Краткое резюме**

Отчёт демонстрирует хорошее понимание принципов ООП и исключительных ситуаций в Java. Все задания выполнены, классы исключений созданы корректно, а методы классов `ArrayTabulatedFunction` и `LinkedListTabulatedFunction` выбрасывают исключения в соответствии с требованиями. Основной плюс – внимательность к деталям и покрытие всех краевых случаев при тестировании. Главный риск – отсутствие исходного кода в отчёте затрудняет оценку качества реализации связного списка и оптимизаций.

**2. Соответствие заданию**

*   **Итог: Да.**
*   Задание 1 выполнено: представлены описания стандартных исключений.
*   Задание 2 выполнено: созданы классы `FunctionPointIndexOutOfBoundsException` и `InappropriateFunctionPointException` с наследованием от соответствующих базовых классов.
*   Задание 3 выполнено: конструкторы и методы `ArrayTabulatedFunction` проверяют условия и выбрасывают исключения. Примеры кода подтверждают это: "В конструкторе", "В методах по индексу", "В setPointX", "В deletePoint".
*   Задание 4 выполнено: реализован класс `LinkedListTabulatedFunction` на основе двусвязного циклического списка.

**3. Техническая корректность (ООП и Java)**

*   Классы исключений (`FunctionPointIndexOutOfBoundsException`, `InappropriateFunctionPointException`) корректно наследуются от базовых классов (`IndexOutOfBoundsException`, `Exception` соответственно) – по описанию.
*   Принципы ООП: Инкапсуляция реализована через приватный статический вложенный класс `FunctionNode` в `LinkedListTabulatedFunction` – по описанию.
*   Использование исключений для обработки ошибок – корректно.
*   В `ArrayTabulatedFunction` используется приватный метод `checkIndex` для повторного использования логики проверки границ – хороший пример.
*   В `LinkedListTabulatedFunction` описаны методы для работы со списком, что соответствует заданию.
*   По описанию, реализована оптимизация доступа к элементам списка в `LinkedListTabulatedFunction`.

**4. Архитектура и дизайн**

*   Структура пакетов соответствует заданию.
*   В `LinkedListTabulatedFunction` использована структура двусвязного циклического списка с выделенной головой, как и требовалось.
*   Обоснование выбора приватного статического класса `FunctionNode` – логично и соответствует принципам инкапсуляции.
*   Реализация интерфейса `TabulatedFunction` позволяет использовать оба класса функций полиморфно.

**5. Воспроизводимость (опционально)**

*   Не представлено — допустимо.

**6. Тестирование и результаты (опционально)**

*   Представлены результаты консольного вывода, демонстрирующие работу тестов для `ArrayTabulatedFunction` и `LinkedListTabulatedFunction`.
*   Покрытие тестами включает проверку исключений и нормальной работы методов.
*   Результаты тестов подтверждают корректность работы реализованных классов.

**7. Стиль и конвенции (применимо при наличии кода)**

*   Код не предоставлен — допустимо.

**8. Источники и добросовестность**

*   Указаны изученные стандартные исключения.
*   Нет видимых заимствований.

**9. Рекомендации по улучшению**

*   **Добавьте исходный код:** Предоставление кода позволит более детально оценить качество реализации связного списка, оптимизации и стиль написания.
*   **Добавьте UML-диаграмму:**  Визуальное представление структуры классов `ArrayTabulatedFunction` и `LinkedListTabulatedFunction` поможет лучше понять архитектуру и связи между ними.
*   **Обоснуйте выбор двусвязного циклического списка:** Укажите, какие преимущества этот тип списка даёт по сравнению с другими структурами (например, односвязным списком).
*   **Продемонстрируйте оптимизацию в коде:** Предоставьте фрагменты кода, иллюстрирующие реализованные оптимизации доступа к элементам списка в `LinkedListTabulatedFunction`.
*   **Добавьте модульные тесты:** Написание модульных тестов (JUnit) позволит более надежно проверить корректность работы классов и избежать регрессий.
*   **Улучшите ясность изложения:** Хотя отчёт в целом понятен, можно добавить больше деталей о сложностях реализации связного списка и принятых решениях.


# Результаты проверки текста отчета на плагиаризм
Обнаружены значительные совпадения с работами HannahMontana2006, в частности, с описанием задания по изучению исключений Java и созданием пользовательских исключений. Также присутствует сходство с кодом, представленным Egor-Snow и sharllottka, включающее определения методов класса.

# Результаты проверки кода
## Рецензия на лабораторную работу

**Работоспособность:**

Код в целом соответствует заданию. Реализованы все требуемые классы исключений (`FunctionPointIndexOutOfBoundsException`, `InappropriateFunctionPointException`) и методы в классах `ArrayTabulatedFunction` и `LinkedListTabulatedFunction` для работы с табулированными функциями.  Конструкторы выбрасывают исключения при некорректных параметрах, методы `getPoint()`, `setPoint()` и другие выбрасывают исключения при выходе за границы массива/списка.  Метод `deletePoint()` корректно выбрасывает исключение при попытке удаления точек, когда их количество меньше 3.  Тесты в `Main.java` покрывают основные сценарии выброса исключений и демонстрируют работоспособность системы.

**Качество кода:**

*   **Читаемость и понятность:** Код достаточно читаемый, структура классов логична. Использование комментариев умеренно, но в целом достаточно для понимания логики. Названия переменных и методов в основном соответствуют их назначению.
*   **Простота и избегание избыточности:** В `LinkedListTabulatedFunction` реализована оптимизация доступа к элементам списка (проход с головы или конца в зависимости от индекса).  В целом, код не содержит явных избыточностей.
*   **Соответствие базовым рекомендациям по стилю:** Стиль кода соответствует общепринятым соглашениям Java.

**Проблемные участки:**

В `LinkedListTabulatedFunction`, при реализации метода `addPoint`, проверка на совпадение x происходит только до добавления нового узла.  Внутри цикла `while`, если `current.data.getX()` равен `newX`, выбрасывается исключение. Однако, после выхода из цикла, необходимо также проверить `current != head`, так как после выхода из цикла `current` может указывать на `head`, и тогда проверка будет лишней.

В целом, код выполнен качественно и соответствует требованиям задания.  Реализованные классы и методы обеспечивают корректную работу с табулированными функциями и обработку исключительных ситуаций.

DiBoS290699 commented:
Расширьте нормальное тестирование класса LinkedList с демонстрацией точек до и после изменений (setPoint, deletePoint, addPoint...)

Dakulzz committed:
Main.java: additional testing (+minor readability)

Before-after testing for [add, set, delete]. 

Minor readability adjustments: 
print() for System.out.println().
printAllPoints for full array console output
pointToString for point's conversion to string

DiBoS290699 reviewed (commented):
# Результаты проверки текста отчета
## Рецензия на отчёт по лабораторной работе №3 (ООП на Java)

**1. Краткое резюме**

Отчёт демонстрирует хорошее понимание поставленной задачи и принципов ООП. Реализованы классы `ArrayTabulatedFunction` и `LinkedListTabulatedFunction` с обработкой исключений в соответствии с заданием. Основной плюс – внимательность к требованиям и корректная реализация исключительных ситуаций. Главный пробел – отсутствие исходного кода в отчёте, что затрудняет детальную проверку реализации.

**2. Соответствие заданию**

*   **Итог: Да**
*   Все задания реализованы:
    *   Классы исключений `FunctionPointIndexOutOfBoundsException` и `InappropriateFunctionPointException` созданы и наследуют от соответствующих базовых классов.
    *   Конструкторы `ArrayTabulatedFunction` выбрасывают исключение `IllegalArgumentException` при некорректных параметрах.
    *   Методы `ArrayTabulatedFunction` выбрасывают исключения в соответствии с требованиями.
    *   Класс `LinkedListTabulatedFunction` реализован с использованием двусвязного циклического списка.
    *   Реализован интерфейс `TabulatedFunction`.
*   В отчёте приводятся примеры выбрасываемых исключений и соответствующие сообщения, что подтверждает соответствие заданию. ("OK getPoint(-1): Индекс -1 вне диапазона [0, 4]")

**3. Техническая корректность (ООП и Java)**

*   **По описанию:**
    *   **Инкапсуляция:** Принцип инкапсуляции реализован в классе `LinkedListTabulatedFunction` путем сокрытия деталей реализации связного списка (вложенный класс `FunctionNode` - приватный).
    *   **Абстракция:** Интерфейс `TabulatedFunction` определяет общие методы, абстрагируясь от конкретной реализации.
    *   **Наследование:** Классы исключений правильно наследуют от базовых классов исключений Java.
    *   **Обработка исключений:** Механизм `try-catch` (в классе `Main`) используется для перехвата исключений.
    *   **Разделение ответственности:**  В `LinkedListTabulatedFunction` наблюдается разделение ответственности между классом списка и классом табулированной функции.
    *   **Композиция:** Отношение композиции между `LinkedListTabulatedFunction` и `FunctionNode`.
*   **По коду (отсутствует):** Невозможно оценить корректность реализации, нейминг, форматирование и Javadoc.

**4. Архитектура и дизайн**

*   **Структура пакетов:** Пакет `functions` содержит классы, связанные с табулированными функциями.
*   **Паттерны:** Отчёт не упоминает использование конкретных паттернов проектирования.
*   **UML-диаграммы:** Отсутствуют.
*   **Обоснование решений:** Обоснование выбора двусвязного циклического списка приведено ("Реализована оптимизация доступа: выбирается ближайший путь: от головы (head.next) или от хвоста (head.prev).").  Обоснование инкапсуляции `FunctionNode` также присутствует ("Элементы списка (FunctionNode) являются внутренней деталью реализации LinkedListTabulatedFunction и не должны быть доступны извне (инкапсуляция).")

**5. Воспроизводимость (опционально)**

*   Не представлено — допустимо.

**6. Тестирование и результаты (опционально)**

*   Приведены результаты тестирования, демонстрирующие корректное выбрасывание исключений в различных сценариях ("OK getPoint(-1): Индекс -1 вне диапазона [0, 4]"). Это говорит о наличии тестов, хотя они и не представлены в коде.
*   Описаны тестовые случаи (некорректные параметры конструктора, выход за границы индекса, некорректное изменение X, добавление точки с существующим X, удаление при минимальном количестве точек, нормальная работа).
*   Интерпретация результатов присутствует.

**7. Стиль и конвенции (применимо при наличии кода)**

*   Код не предоставлен — допустимо.

**8. Источники и добросовестность**

*   Указаны изученные классы исключений Java.
*   Корректные ссылки на внешние материалы отсутствуют.
*   Подозрение на плагиат не обнаружено.

**9. Рекомендации по улучшению**

*   **Добавьте исходный код:** Предоставление кода позволит более детально оценить качество реализации, нейминг, форматирование и соответствие Java-конвенциям.
*   **Добавьте UML-диаграммы:** Диаграммы классов помогут визуализировать архитектуру и связи между классами, что улучшит понимание дизайна.
*   **Обоснуйте выбор паттернов (если использовались):** Если в реализации были использованы паттерны проектирования, опишите их и объясните, почему они были выбраны.
*   **Представьте модульные тесты:** Добавление JUnit-тестов с проверкой граничных случаев, негативных сценариев и покрытия кода повысит надежность и качество работы.
*   **Добавьте раздел воспроизводимости:** Опишите инструменты, версии, команды сборки и запуска тестов, чтобы другие разработчики могли легко воспроизвести результаты.
*   **Улучшите ясность изложения:** Укажите конкретные примеры входных данных и ожидаемых результатов при тестировании.


# Результаты проверки текста отчета на плагиаризм
Обнаружены значительные совпадения с работами HannahMontana2006, касающиеся изучения стандартных исключений в Java и создания пользовательских исключений. Также присутствует сходство с кодом, представленным Egor-Snow и sharllottka, в части объявлений методов интерфейса.

# Результаты проверки кода
## Рецензия лабораторной работы

### Работоспособность

В целом, код соответствует поставленной задаче и требованиям лабораторной работы. Реализованы все необходимые классы исключений и методы для работы с табулированными функциями.  Конструкторы выбрасывают исключения `IllegalArgumentException` при некорректных параметрах, методы `getPoint()`, `setPoint()`, `getPointX()`, `setPointY()` и `deletePoint()` выбрасывают исключения `FunctionPointIndexOutOfBoundsException` при выходе за границы набора точек, а `setPoint()` и `addPoint()` выбрасывают исключения `InappropriateFunctionPointException` при попытке добавить или изменить точку функции несоответствующим образом. Метод `deletePoint()` выбрасывает исключение `IllegalStateException`, если на момент удаления точки количество точек в наборе меньше трех.

Однако, есть замечания по корректности работы интерполяции, особенно при граничных значениях, и реализации проверки x на существование, что было обнаружено при тестировании.

### Качество кода

Качество кода в целом хорошее, но есть возможности для улучшения:

*   **Читаемость и понятность:** Названия переменных и методов достаточно информативны. Комментарии присутствуют и поясняют логику работы. Структура классов логична и соответствует объектно-ориентированному подходу.
*   **Простота и избегание избыточности:** В коде присутствует некоторая избыточность, например, дублирование логики проверки индексов в нескольких методах.  Можно выделить общую функцию для проверки индексов.
*   **Соответствие базовым рекомендациям по стилю:** В целом, код соответствует базовым рекомендациям по стилю Java. Однако, стоит обратить внимание на константу EPSILON, возможно, ее стоит сделать final static.

**Пример проблемного участка кода (LinkedListTabulatedFunction.java):**

В методе `addPoint()` логика проверки на наличие точки с таким же x реализована неоптимально. Перебор списка до вставки неэффективен, особенно для больших списков.  Стоит рассмотреть использование более эффективных структур данных, например, `HashSet`, для хранения абсцисс точек.

```java
        // проверка на совпадение x
        for (int i=0;i<pointsCount;i++) {
            if (Math.abs(point.getX() - points[i].getX()) < EPSILON) {
                throw new InappropriateFunctionPointException("Точка с таким x уже существует");
            }
        }
```

**Рекомендации:**

*   Выделить общую функцию для проверки индексов.
*   Оптимизировать поиск точки с заданным x в методе `addPoint()`.
*   Протестировать интерполяцию на граничных значениях.
*   Рассмотреть возможность использования `HashSet` для хранения абсцисс точек.
*   Проанализировать и оптимизировать производительность методов `addNodeByIndex()` и `deleteNodeByIndex()`.
*   Добавить Javadoc для всех классов и методов.

DiBoS290699 commented:
Принято, но не забывайте всегда обновлять отчёт
